### DISCUSSION TITLE
Different history search behavior when vi-mode is active

### DISCUSSION THREAD

**Discussion by S1ngS1ng** [ohmyzsh/ohmyzsh#11970] [Help]
Here is how it works when `vi-mode` is NOT active:
![2019-04-28_11-46-29 (1)](https://user-images.githubusercontent.com/13592559/56868794-379f7800-69ac-11e9-872f-fe7bd2df1cee.gif)
This is my expected behavior where completion is performed based on the input (a.k.a. "expand" completion result)

While `vi-mode` is active:
![2019-04-28_11-49-06 (1)](https://user-images.githubusercontent.com/13592559/56868814-6e758e00-69ac-11e9-9b2f-99803a4f0867.gif)
The completion does not take the input. Instead, the completion works as if nothing was typed in.

Little bit more details if anyone is interested:
1. I am using the arrow key `up` for completion
2. The only difference in my `.zshrc` is under `plugins`:
```diff
plugins=(
  git
  zsh-syntax-highlighting
  zsh-autosuggestions
+ vi-mode
)
```

**Comment by mcornella** [comment]
Try putting `vi-mode` first on the list of plugins. This is an issue with the order in which bindkey modes are activated.

**Comment by mcornella** [comment]
Closing since this is in essence a duplicate of #3435.

**Comment by S1ngS1ng** [comment]
@mcornella Hey Marc, thanks for your reply.
I've tried as you suggested, still it doesn't fix the issue.
Also, just F.Y.I., I am not using the `history-substring-search` plugin.

Here is some information that might help:
1. The partial output of `bindkey` when `vi-mode` **IS NOT** in `plugins`:
![image](https://user-images.githubusercontent.com/13592559/56994909-3e1d2380-6b55-11e9-8bf7-079116d80f54.png)
2. The partial output of `bindkey` when `vi-mode` **IS** in `plugins`, **no matter where it is**:
![image](https://user-images.githubusercontent.com/13592559/56995004-73c20c80-6b55-11e9-87d1-b0df68fb8bb7.png)
3. Here is the output of `up` after `control + shift + v`
![image](https://user-images.githubusercontent.com/13592559/56995070-8e948100-6b55-11e9-9245-f8e8834d5a79.png)
4. I am on macOS. The terminal emulator that I use is iTerm2

So, I am wondering if the behavior I mentioned above was due to the different method being called? If so, how may I override that?
If you need any other info such as debug output, please let me know.

**Comment by mcornella** [comment]
Ok, then this is the effect of switching to viins keymap while those key bindings are put on the emacs keymap—from `lib/key-bindings.zsh` file: https://github.com/robbyrussell/oh-my-zsh/blob/4eae0549fd7e8a6d35c5627356098380b6b9ef82/lib/key-bindings.zsh#L31-L41

#5113 should be able to solve this, but in the meantime, add the correct bindkey commands to your zshrc file, after OMZ is sourced.

**Comment by sprytnyk** [comment]
I was interested in the same thing and solution to this. I added to the end of my `.zshrc` the following lines and it did a trick for me but it may be absolutely wrong way of doing this but still, it works but in `insert` mode, `normal` one keeps the old behaviour, so for me, it's even better to combine the best of two worlds:
```sh
$ cat >> ~/.zshrc <<EOF
bindkey "^P" up-line-or-history
bindkey "^[OA" up-line-or-beginning-search
bindkey "^[[5~" up-line-or-history
bindkey "^[[A" up-line-or-history
bindkey "^N" down-line-or-history
bindkey "^[OB" down-line-or-beginning-search
bindkey "^[[6~" down-line-or-history
bindkey "^[[B" down-line-or-history
EOF
```

**Comment by vtronko** [comment]
@vald-phoenix you are the MVP, sir, thank you.
